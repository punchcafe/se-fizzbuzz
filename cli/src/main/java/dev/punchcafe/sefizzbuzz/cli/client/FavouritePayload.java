package dev.punchcafe.sefizzbuzz.cli.client;

import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;

@AllArgsConstructor
@Getter
@EqualsAndHashCode
public class FavouritePayload {

    public static FavouritePayload buildFrom(final boolean isFavourite){
        return new FavouritePayload(isFavourite);
    }

    private final boolean is_favourite;
}
