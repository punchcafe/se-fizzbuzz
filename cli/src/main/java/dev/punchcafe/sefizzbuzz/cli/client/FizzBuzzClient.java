package dev.punchcafe.sefizzbuzz.cli.client;

import feign.Param;
import feign.RequestLine;

public interface FizzBuzzClient {

    @RequestLine("GET /fizzbuzz/{fizzBuzzId}")
    FizzBuzzEntity calculate(@Param("fizzBuzzId") int fizzBuzzId);
}
