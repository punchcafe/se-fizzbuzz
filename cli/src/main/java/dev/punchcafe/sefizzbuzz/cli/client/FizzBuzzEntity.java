package dev.punchcafe.sefizzbuzz.cli.client;

import lombok.AllArgsConstructor;
import lombok.Value;

@Value
@AllArgsConstructor
public class FizzBuzzEntity {
    private final int id;
    private final String value;
    private final boolean isFavourite;
}
