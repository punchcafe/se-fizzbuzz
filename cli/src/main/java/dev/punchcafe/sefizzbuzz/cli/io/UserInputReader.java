package dev.punchcafe.sefizzbuzz.cli.io;

import java.util.Scanner;

public class UserInputReader {

    private final Scanner userInput = new Scanner(System.in);

    public String getUserInput(){
        return userInput.nextLine();
    }
}
