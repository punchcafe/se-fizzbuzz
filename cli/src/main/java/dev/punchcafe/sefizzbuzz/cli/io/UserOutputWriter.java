package dev.punchcafe.sefizzbuzz.cli.io;

public class UserOutputWriter {

    public void printToUser(final String output){
        System.out.println(output);
    }
}
