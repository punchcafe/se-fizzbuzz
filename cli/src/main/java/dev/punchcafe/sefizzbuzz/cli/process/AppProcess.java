package dev.punchcafe.sefizzbuzz.cli.process;

import java.util.List;

public interface AppProcess {

    String getProcessName();

    void execute(List<String> args);

}
