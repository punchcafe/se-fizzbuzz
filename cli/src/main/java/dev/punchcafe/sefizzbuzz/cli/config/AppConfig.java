package dev.punchcafe.sefizzbuzz.cli.config;

import dev.punchcafe.sefizzbuzz.cli.client.FizzBuzzClient;
import dev.punchcafe.sefizzbuzz.cli.io.UserOutputWriter;
import lombok.Builder;
import lombok.Value;

@Value
@Builder
public class AppConfig {
    private final UserOutputWriter userOutputWriter;
    private final FizzBuzzClient fizzBuzzClient;
}
