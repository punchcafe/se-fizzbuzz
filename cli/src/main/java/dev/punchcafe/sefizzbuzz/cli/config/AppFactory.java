package dev.punchcafe.sefizzbuzz.cli.config;

import dev.punchcafe.sefizzbuzz.cli.process.AppProcess;
import dev.punchcafe.sefizzbuzz.cli.process.Help;
import dev.punchcafe.sefizzbuzz.cli.process.OptionSelector;
import lombok.AllArgsConstructor;

import java.util.List;

@AllArgsConstructor
public class AppFactory {

    private final AppConfig config;

    public AppProcess buildApp(){
        final var helpCommand = new Help(config.getUserOutputWriter());
        final var initialSelector = OptionSelector.builder()
                .processName("main")
                .subAppProcesses(List.of(helpCommand)).build();
        return initialSelector;
    }
}
