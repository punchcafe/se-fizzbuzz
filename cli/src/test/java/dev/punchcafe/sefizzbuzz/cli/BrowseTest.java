package dev.punchcafe.sefizzbuzz.cli;

import dev.punchcafe.sefizzbuzz.cli.client.FizzBuzzClient;
import dev.punchcafe.sefizzbuzz.cli.config.AppConfig;
import dev.punchcafe.sefizzbuzz.cli.config.AppFactory;
import dev.punchcafe.sefizzbuzz.cli.io.UserInputReader;
import dev.punchcafe.sefizzbuzz.cli.io.UserOutputWriter;
import dev.punchcafe.sefizzbuzz.cli.process.AppProcess;
import dev.punchcafe.sefizzbuzz.cli.snapshot.SnapshotConfig;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;

import java.util.List;

import static dev.punchcafe.sefizzbuzz.cli.TestConstants.BROWSE_FIRST_PAGE_RESPONSE;
import static io.github.jsonSnapshot.SnapshotMatcher.*;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.*;

public class BrowseTest {


    private FizzBuzzClient fizzBuzzClient;
    private UserOutputWriter userOutputWriter;
    private UserInputReader userInputReader;
    private AppProcess appProcess;

    @BeforeAll
    static void beforeAll(){
        start(SnapshotConfig::asJsonString);
    }

    @AfterAll
    static void afterAll(){
        validateSnapshots();
    }

    @BeforeEach
    void beforeEach() {
        userOutputWriter = Mockito.spy(new UserOutputWriter());
        userInputReader = Mockito.mock(UserInputReader.class);
        fizzBuzzClient = Mockito.mock(FizzBuzzClient.class);
        final var appConfig = AppConfig.builder()
                .userOutputWriter(userOutputWriter)
                .userInputReader(userInputReader)
                .fizzBuzzClient(fizzBuzzClient)
                .build();
        appProcess = new AppFactory(appConfig).buildApp();
    }

    @Test
    void userCanBrowse() {
        // set up expected view
        when(fizzBuzzClient.getPage(eq(5), eq(1)))
                .thenReturn(BROWSE_FIRST_PAGE_RESPONSE);
        // set up user command
        when(userInputReader.getUserInput())
                .thenReturn("exit");
        appProcess.execute(List.of("browse"));
        verify(fizzBuzzClient, times(1)).getPage(5, 1);
        final var userOutputCaptor = ArgumentCaptor.forClass(String.class);
        verify(userOutputWriter, times(1))
                .printToUser(userOutputCaptor.capture());
        expect(userOutputCaptor.getValue()).toMatchSnapshot();
    }
}
