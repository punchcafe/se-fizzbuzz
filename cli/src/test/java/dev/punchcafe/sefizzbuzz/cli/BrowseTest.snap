dev.punchcafe.sefizzbuzz.cli.BrowseTest.userCanBrowse=[
  "\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nPAGE     : 1Displaying : 5entities per page \n1 . 1                     2 . 2                     3 . Fizz                  \n4 . 4                     5 . Buzz                  \n enter help for list of instructions"
]